### PR TITLE
ddl: fix alter table charset bug that change blob column type to text

### PR DESCRIPTION
**Pull request by crazycs520**
### What problem does this PR solve? 
Alter table charset will change `blob` column to `text` column. just like below.

Before
```sql
tidb> create table t1 (a blob) character set utf8;
Query OK, 0 rows affected
Time: 0.005s
tidb> alter table t1 charset=utf8mb4 collate=utf8mb4_bin;
Query OK, 0 rows affected
Time: 0.005s
tidb> show create table t1
+-------+-------------------------------------------------------------+
| Table | Create Table                                                |
+-------+-------------------------------------------------------------+
| t1    | CREATE TABLE `t1` (                                         |
|       |   `a` text DEFAULT NULL                                     |
|       | ) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin |
+-------+-------------------------------------------------------------+

```

### What is changed and how it works?
* fix Alter table column bug.

### Check List 

Tests 

 - Unit test
 
Code changes

 - Has exported function/method change

Side effects

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: ddl/table.go [hunk 1]
```diff
@@ -735,6 +735,9 @@ func onModifyTableCharsetAndCollate(t *meta.Meta, job *model.Job) (ver int64, _
 	tblInfo.Collate = toCollate
 	// update column charset.
 	for _, col := range tblInfo.Columns {
+		if col.Charset == charset.CharsetBin {
```
Comment (winkyao) [inline ddl/table.go]: We can use `HasCharset` here.

File: ddl/table.go [hunk 2]
```diff
@@ -735,7 +736,7 @@ func onModifyTableCharsetAndCollate(t *meta.Meta, job *model.Job) (ver int64, _
 	tblInfo.Collate = toCollate
 	// update column charset.
 	for _, col := range tblInfo.Columns {
-		if typesNeedCharset(col.Tp) {
+		if field_types.HasCharset(&col.FieldType) {
```
Comment (bb7133) [inline ddl/table.go]: I guess `typesNeedCharset` can be replaced by `HasCharset`. Another code that calls this function is:

https://github.com/pingcap/tidb/blob/5c21fa4b4345ef61d8ed936c86d3ee477fcf24f6/ddl/ddl_api.go#L299

Can you confirm it here?

Comment (crazycs520) [reply ddl/table.go]: No, https://github.com/pingcap/tidb/blob/5c21fa4b4345ef61d8ed936c86d3ee477fcf24f6/ddl/ddl_api.go#L299 here may be have not set the column Flag, use `HasCharset ` here may be got some error.

Comment (winkyao) [reply ddl/table.go]: so why you use `HashCharset` finally?

Comment (crazycs520) [reply ddl/table.go]: `HashCharset ` can be used here(`ddl/table.go #739`), but `ddl/ddl_api.go #299` should use `typesNeedCharset` function.

Comment (zimulala) [reply ddl/table.go]: Sorry, I still don't quite understand, why is it correct to change to `HasCharset`?

**Pull request merged**